### ISSUE TITLE
RC-17 import from v1 issues

### ISSUE THREAD

**Issue by Cylix** [Requarks/wiki#1174]
**Describe the bug**
Thanks for the hard work on the v2, it looks awesome and I can't wait to get it working and replace our current v1 installation.

I attempted to set up the v2 today, but I came across two issues after importing the data from v1 using the `utilities` tool in the admin area.

The first one involves broken links.
The reason is that for pages created with the v1 and containing underscores `_`, the actual markdown file created did not contain these underscores (for example, creating a page `/weekly_summary` would be stored as `/weeklysummary.md`). This worked still fine on v1, because accessing either `/weekly_summary` or `/weeklysummary` from the browser would lead to the same `/weeklysummary.md`. However, the v2 seems to make a distinction between paths containing underscores and those who don't, and it basically breaks the links we got from the v1. This issue is inconvenient, but we can get to solve it by manually fixing the links. Just thought it could be good to know on your end.

The second issue is much more critical though. Some pages imported from the v1 are blanked when watching them from the browser. I checked, and the content is successfully imported locally though (the file is located under `/data/repo` and contains the whole content). If I attempt to look at the history from the admin area, or even to open the edit page from the edit area, I can see the content and even the preview in the editor. But opening the page just leads to a blank page. There is no error log when running `sudo journalctl -u wiki` as well.

I put the content of on of the page below so that you can try to reproduce on your end. Created the page with that exact content reproduces the error on my end. I'll try to track down what's creating this issue.

https://pastebin.com/dc6EaqBZ

Additionally, I got one page that got duplicated. That is, when I looked on the admin area in the page section and filtered by path, I saw one page appearing twice with the same content. I originally thought it was the root cause of the blank page because I found that while investigating the issue, but deleting the duplicate did not solve the issue, and I eventually came across other blank pages without duplicates (so the blank page is purely based on the content, the duplicate appears to be a separate issue)

**Screenshots**
The issue seems to be JS-related.

When looking for console errors, I got:
<img width="1917" alt="Screen Shot 2019-11-02 at 4 20 13 PM" src="https://user-images.githubusercontent.com/3252462/68078086-cd1d2980-fd8c-11e9-9d2f-ede07be9c175.png">

When looking the content of the body with  the chrome inspector, it is empty:
<img width="1595" alt="Screen Shot 2019-11-02 at 4 20 26 PM" src="https://user-images.githubusercontent.com/3252462/68078091-d7d7be80-fd8c-11e9-99a2-4cce1bf31c28.png">

However, oddly enough, when checking the source code with chrome `view-source`, the content of the page is loaded successfully:
<img width="1755" alt="Screen Shot 2019-11-02 at 4 20 39 PM" src="https://user-images.githubusercontent.com/3252462/68078103-efaf4280-fd8c-11e9-8216-18d268fbda2b.png">

**Host Info (please complete the following information):**
I'm running the latest RC version on ubuntu 18.04 LTS.

We are running wiki.js on top of a MySQL `5.7.14-google-log` (GCloud mysql instance).

**Additional context**
I tried to import the v1 content twice from a fresh v2 install with fresh mysql DB, both times lead to the same result. Each time, I waited until the import fully finished before even browsing a single page.

**Comment by Cylix** [comment]
OK, tracked down the blank page from `{true|false}` content.

Just creating a page with that content and nothing else reproduces the bug.

It results in the following HTML code `<div v-pre><p true|false></p></div>` which is invalid. It works well on v1 though.

I'll retry the import from scratch and try to solve these issues by manually editing and see if I come across more issues.

**Comment by Cylix** [comment]
Another minor issue when importing.

On v1, titles & subtitles where setup from the `TITLE` & `SUBTITLE` comments. Seems there is a glitch in the v1 import script that breaks when there is missing spaces in the closing of the comments, leading to thte title/subtitle to not be imported correctly:

<img width="845" alt="Screen Shot 2019-11-02 at 4 53 38 PM" src="https://user-images.githubusercontent.com/3252462/68078359-b9c08d00-fd91-11e9-9fd7-41cb4bbc36bf.png">
<img width="466" alt="Screen Shot 2019-11-02 at 4 53 45 PM" src="https://user-images.githubusercontent.com/3252462/68078360-b9c08d00-fd91-11e9-8e84-8e240e57f450.png">

The first screenshot is the import title without subtitle.

The second screenshot is the original content from v1. Note the missing space before `-->` for the title

**Comment by Cylix** [comment]
Similar issue as the broken links due to underscores in the path, we have broken links for links containing dots in the path (for example `/releases/6.6.0` was stored as `/release/660.md`. Accessing `/releases/6.6.0` would work in v1, not v2), will manually patch the links for now

**Comment by Cylix** [comment]
Minor issue, not sure if that's a bug or if I didn't get the feature.

For links in the v2, some are displayed in blue, some in reds. I thought red meant that the link was broken, but some links are red but are not broken, so not sure of the root cause:

<img width="269" alt="Screen Shot 2019-11-02 at 5 08 09 PM" src="https://user-images.githubusercontent.com/3252462/68078475-76671e00-fd93-11e9-93dd-f223699821cc.png">

The first red link is `/public/ios/build`.
The second is `/public/ios/push`.
Both work fine

**Comment by Cylix** [comment]
For a very large page, it failed to be imported and I can't create it manually. Got the following error:

```
insert into `pages` (`authorId`, `content`, `contentType`, `createdAt`, `creatorId`, `description`, `editorKey`, `hash`, `isPrivate`, `isPublished`, `localeCode`, `path`, `publishEndDate`, `publishStartDate`, `title`, `toc`, `updatedAt`) values (REDACTED) - Data too long for column 'content' at row 1
```

**Comment by NGPixel** [comment]
- The paths are based on the filenames. If you used underscores or dots in your links v1, these won't work in v2. You'll need to rename the pages to use dashes / underscore. Note that dots are not allowed in paths in v2.

- The v1 TITLE and SUBTITLE not working without a space will be fixed in the next build.

- Blue/Red Links Mismatch: Try to edit the page containing the links and save again. Links will be reconnected.

I'm looking at the sample page content you provided. Using { } outside of code blocks is likely to cause rendering issue at the moment.

**Comment by Cylix** [comment]
`Blue/Red Links Mismatch: Try to edit the page containing the links and save again. Links will be reconnected.`

I tried this one, but it does not work

I found the root cause though. The link in the markdown included a trailing slash (e.g.: `/my/path/`). Removing the trailing slash fixes the issue.

**Comment by Cylix** [comment]
I think there is a problem regarding permissions.

Let's say I enable all the global content permissions for one group. Then I do the following:
- enable read:pages for /
- disable manage:pages & write:pages for /

Then, if I login with one user of that group, I can't read anything under /, so seems the deny overrides the allow rule completely, even though I did not deny explicitly read:pages.

If I just allow read:pages under /, I can still edit pages of /, probably because the write:pages & manage:pages global content permission is enabled, but if I disable it I have no way to explicitly enable edit for one page.

Another issue related to permission is that for some changes (don't recall which ones), I had to signout and signin again to see the change on the user side

**Comment by Cylix** [comment]
For the `Another issue related to permission is that for some changes (don't recall which ones), I had to signout and signin again to see the change on the user side`, I found a way to reproduce. Create two groups with different set of permission and login with one user assign to one of these groups in incognito mode. Swith the user to the other group and check back from the user-side: he can still navigate from the first group permission and has to signout and signin again to get the new group permission.

**Comment by Cylix** [comment]
I think that should be it for the issues I noticed, sorry for the multiple messages. Overall, the v2 really looks great and I'm gonna move forward for our company with this RC-17.

I think the major issue right now is about the permissions. I made it partially work for our use case by enabling all read/write permissions for all pages except one specific path, but optimal solution would be enabling read everywhere and write only for some places.

If the content limit per page could be increased, that'd be great too. I fixed one critical page by breaking it down into multiple pages because we were planning to do so either way. But we got some other pages where we copy paste some operation logs that may be quite long. On v1, that worked fine, but on v2 we basically lost access to these pages. This is not critical for now because we don't need these logs for now and we can still access them through the synced git repo either way.

For the other issues mentioned (red links, broken links), I manually fixed it. That may be a problem for people with larger documentation (we have approx. 300 docs pages), so I guess if the import script can detect these links and fix them that'd be awesome for others. Otherwise a warning in the documentation to let them know about the possible broken link is fair enough.

Will let you know if our team come across other issues next week when they will start using the v2. Thanks a lot for the hardwork, this version is very promising!

**Comment by NGPixel** [comment]
Content limit (only occurs when using mysql/mariadb) was fixed by 8eddc4799e134dc568dbf942559df56d4bee8eb4

**Comment by Cylix** [comment]
Thanks! When upgrading to the next version, will this change be applied to the current DB schema?

**Comment by NGPixel** [comment]
Yes, the columns will be changed to LONGTEXT automatically.

**Comment by NGPixel** [comment]
The `{true|false}` blank page bug is now fixed (fb0c64a07e9fceb457dd1c6f19896bcc73d7d77b). Note that you will need to put such content as inline code, as it will now be removed from the rendered content. `{ }` are reserved for assigning classes and IDs to content in v2.
